LEGO1/tgl/tgl.h

```diff
@@ -0,0 +1,363 @@
+#endif
+};
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// View
+
+class View : public Object {
+public:
+	virtual Result Add(const Light*) = 0;
+	virtual Result Remove(const Light*) = 0;
+
+	virtual Result SetCamera(const Camera*) = 0;
+	virtual Result SetProjection(ProjectionType) = 0;
+	virtual Result SetFrustrum(double frontClippingDistance, double backClippingDistance, double degrees) = 0;
+	virtual Result SetBackgroundColor(double r, double g, double b) = 0;
+
+	virtual Result Clear() = 0;
+	virtual Result Render(const Group*) = 0;
+	// ??? needed for fine grain control when using DirectDraw/D3D ???
+	virtual Result ForceUpdate(unsigned long x, unsigned long y, unsigned long width, unsigned long height) = 0;
+
+	// ??? for now: used by Mesh Cost calculation
+	virtual Result TransformWorldToScreen(const double world[3], double screen[4]) = 0;
+
+	// Pick():
+	//  x, y:
+	//      view coordinates
+	//
+	//  ppGroupsToPickFrom:
+	//      array of (Group*) in any order
+	//      Groups to pick from
+	//
+	//  groupsToPickFromCount:
+	//      size of ppGroupsToPickFrom
+	//
+	//  rppPickedGroups:
+	//      output parameter
+	//      array of (Group*) representing a Group hierarchy
+	//      top-down order (element 0 is root/scene)
+	//      caller must deallocate array
+	//      ref count of each element (Group*) has not been increased
+	//      an element will be 0, if a corresponding Group was not found in ppGroupsToPickFrom
+	//
+	//  rPickedGroupCount:
+	//      output parameter
+	//      size of rppPickedGroups
+	virtual Result Pick(
+		unsigned long x,
+		unsigned long y,
+		const Group** ppGroupsToPickFrom,
+		int groupsToPickFromCount,
+		const Group**& rppPickedGroups,
+		int& rPickedGroupCount
+	) = 0;
+};
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// Camera
+
+class Camera : public Object {
+public:
+#if 0
+    virtual Result  SetPosition(const double[3]) = 0;
+    virtual Result  SetOrientation(const double direction[3],
+                                   const double up[3]) = 0;
+#endif
+	virtual Result SetTransformation(const FloatMatrix4&) = 0;
+};
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// Light
+
+class Light : public Object {
+public:
+#if 0
+    virtual Result  SetPosition(const double[3]) = 0;
+    virtual Result  SetOrientation(const double direction[3],
+                                   const double up[3]) = 0;
+#endif
+	virtual Result SetTransformation(const FloatMatrix4&) = 0;
+};
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// Group
+
+class Group : public Object {
+public:
+#if 0
+    virtual Result  SetPosition(const double[3]) = 0;
+    virtual Result  SetOrientation(const double direction[3],
+                                   const double up[3]) = 0;
+#endif
+	virtual Result SetTransformation(const Matrix4&) = 0;
```

Comment:
We should probably add a major TODO comment considering we changed the parameter type here. Seems highly unlikely this would suddenly take a `Matrix4` and not a TGL vector like the others. I know this was changed to make the `UpdateWorldData` code match but it's an odd state to be in, so I have a feeling we might be missing something here.

Reply:
Yes, it should be a FloatMatrix4, as per the commented out declaration in UpdateWorldData. I can't remember now what the issue I had was with that.  I think it generated extra code.